doc/feature/synapses.md

```diff
@@ -0,0 +1,189 @@
+Synapse support in Brain
+============
+
+The low-level data access for synapses provided by Brion needs a complementary
+API in Brain to support various common higher-level uses cases, like semantic
+synapse object access or full-circuit synapse density voxelization, to name the
+two extremes.
+
+## Requirements
+
+* Semantic synapse objects that give read-access to all properties
+* Array access which allows vectorization
+* Filtering of synapses for pathway use case (shared synapses of two GIDSets)
+* Lazy-loading to support out-of-core scenarios
+* User-transparent KVS caching
+
+## New dependencies
+
+None
+
+## API
+
+```cpp
+
+class Circuit
+{
+    SynapsesStream getAfferentSynapses( const GIDSet& gids ) const;
+    SynapsesStream getEfferentSynapses( const GIDSet& gids ) const;
+    SynapsesStream getSharedSynapses( const GIDSet& preGIDs,
+                                      const GIDSet& postGIDs ) const;
+
+    // convenience for upper getXXX( gids ).read( gids.size( )).get()
+    Synapses getAllAfferentSynapses( const GIDSet& gids ) const;
+    Synapses getAllEfferentSynapses( const GIDSet& gids ) const;
+    Synapses getAllSharedSynapses( const GIDSet& preGIDs,
+                                   const GIDSet& postGIDs ) const;
+};
+
+// read-only, moveable proxy object of data storage in Synapses container
+class Synapse
+{
+    uint32_t preGID() const;
+    uint32_t preSectionID() const;
+    uint32_t preSegmentID() const;
+    float preDistance() const;
+    Vector3f preSurfacePosition() const;
+    Vector3f preCenterPosition() const;
+
+    uint32_t postGID() const;
+    uint32_t postSectionID() const;
+    uint32_t postSegmentID() const;
+    float postDistance() const;
+    Vector3f postSurfacePosition() const;
+    Vector3f postCenterPosition() const;
+
+    float conductance() const;
+    float utilization() const;
+    float depression() const;
+    float facilitation() const;
+    float decay() const;
+    int efficacy() const;
+};
+
+// read-only, moveable, non-copyable, contains all data of all synapses
+class Synapses
+{
```

Comment:
Why not adding a conversion constructor (i.e. without explicit) for this class that takes a SynapseStream as input. Then the getAllXXX methods are not needed and one could write:

     Synapses s = circuit.getAfferentSynapses(cells)
     SynapseStream s = circuit.getAfferentSynapses(cells);